apiserver/facades/client/charms/charmhubrepo_test.go

```diff
@@ -4,180 +4,506 @@
-	cp := channelPlatform{
-		Platform: corecharm.MustParsePlatform("amd64"),
+func (s *charmHubRepositoriesSuite) setupMocks(c *gc.C) *gomock.Controller {
+	ctrl := gomock.NewController(c)
+	s.client = mocks.NewMockCharmHubClient(ctrl)
+	return ctrl
+}
+
+func (s *charmHubRepositoriesSuite) expectedCURL(curl *charm.URL, revision int, arch string, series string) *charm.URL {
+	return curl.WithRevision(revision).WithArchitecture(arch).WithSeries(series)
+}
+
+// RefreshConfgMatcher is required so that we do check somethings going into
+// the refresh method. As instanceKey is private we don't know what it is until
+// it's called.
+type RefreshConfgMatcher struct {
+	c      *gc.C
+	Config charmhub.RefreshConfig
+}
+
+func (m RefreshConfgMatcher) Matches(x interface{}) bool {
+	rc, ok := x.(charmhub.RefreshConfig)
+	if !ok {
+		return false
+	}
+
+	cb, _, err := m.Config.Build()
+	m.c.Assert(err, jc.ErrorIsNil)
+
+	rcb, _, err := rc.Build()
+	m.c.Assert(err, jc.ErrorIsNil)
+	m.c.Assert(len(cb.Actions), gc.Equals, len(rcb.Actions))
+
+	return cb.Actions[0].ID == rcb.Actions[0].ID
+}
+
+func (RefreshConfgMatcher) String() string {
+	return "is refresh config"
+}
+
+type refreshConfigSuite struct {
+	testing.IsolationSuite
+}
+
+var _ = gc.Suite(&refreshConfigSuite{})
+
+func (refreshConfigSuite) TestRefreshByChannel(c *gc.C) {
+	curl := charm.MustParseURL("ch:wordpress")
+	platform := corecharm.MustParsePlatform("amd64/ubuntu/focal")
+	channel := corecharm.MustParseChannel("latest/stable").Normalize()
+	origin := corecharm.Origin{
+		Platform: platform,
+		Channel:  &channel,
 	}
 
-	override, matched := cp.MatchPlatform(transport.InfoChannelMap{
-		Channel: transport.Channel{
-			Platform: transport.Platform{
+	cfg, err := refreshConfig(curl, origin)
+	c.Assert(err, jc.ErrorIsNil)
+
+	ch := channel.String()
+	instanceKey := charmhub.ExtractConfigInstanceKey(cfg)
+
+	build, _, err := cfg.Build()
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(build, gc.DeepEquals, transport.RefreshRequest{
+		Actions: []transport.RefreshRequestAction{{
+			Action:      "install",
+			InstanceKey: instanceKey,
+			Name:        &curl.Name,
+			Channel:     &ch,
+			Platform: &transport.RefreshRequestPlatform{
+				OS:           "ubuntu",
+				Series:       "focal",
 				Architecture: "amd64",
 			},
-		},
+		}},
+		Context: []transport.RefreshRequestContext{},
 	})
-	c.Assert(override.Arch, jc.IsFalse)
-	c.Assert(override.Series, jc.IsFalse)
-	c.Assert(matched, jc.IsTrue)
 }
 
-func (s channelPlatformSuite) TestMatchArchAll(c *gc.C) {
-	cp := channelPlatform{
-		Platform: corecharm.MustParsePlatform("amd64"),
+func (refreshConfigSuite) TestRefreshByRevision(c *gc.C) {
+	revision := 1
+	curl := charm.MustParseURL("ch:wordpress")
+	platform := corecharm.MustParsePlatform("amd64/ubuntu/focal")
+	channel := corecharm.MustParseChannel("latest/stable").Normalize()
+	origin := corecharm.Origin{
+		Platform: platform,
+		Channel:  &channel,
```

Comment:
test has inacurate expectations, calls to refresh will fail if both channel and revision are provided:
 "error": {
        "code": "revision-conflict",
        "message": "Both revision and channel are present in the request for this Charm. It should be one or the other exclusively."
      },